src/IB/IBFEMethod.cpp

```diff
@@ -988,7 +988,8 @@ IBFEMethod::interpolateVelocity(const int u_data_idx,
                                                  /*close_F*/ false,
                                                  /*close_X*/ false);
     }
-    batch_vec_assembly(d_U_rhs_vecs);
+    batch_vec_ghost_update(d_U_rhs_vecs, ADD_VALUES, SCATTER_REVERSE);
```

Comment:
Do we need to have a switch depending on whether the vector is `GHOSTED` or not?

Reply:
Yup. I will fix this.

I also need to add some more tests that verify things work with both ghosted and unghosted vectors.

Reply:
> I also need to add some more tests that verify things work with both ghosted and unghosted vectors.

Does it make sense for there to be a flag in `IBFEMethod` that determines the vector type?

Reply:
Hm. IBFEMethod already passes ghosted vectors to this function, but CIBFEMethod does not. I would prefer to use ghosted vectors internally for this type of assembly.

What would this flag do? I am not sure that I understand your proposal.

Reply:
Sure: The flag would determine if we use ghosted or non-ghosted vectors in this interpolation routine. Right now we always use ghosted vectors in `IBFEMethod` for this stuff, but it should be easy enough to switch to non-ghosted. This would make it easy to measure the effect of switching from using `VecSetValues` and the internal stash, to using fully local assembly followed by a scatter.

(Note that the coordinate vectors will always have to be ghosted so that we know where the interaction points are.)

Reply:
:+1:

Reply:
To try to avoid confusion: My proposal is that in `IBFEMethod`, we would dynamically determine the type of `d_U_IB_rhs_vecs` when we set those vectors up.

This could also help to test/verify the implementation is doing the right thing.

Reply:
I think that we still need to have some logic here to decide how to assemble, and also possibly in `IBFEMethod` and other places where we batch assemble following calls to `interpWeighted()`.

Reply:
Absolutely. In fact, the new test that does not use ghosted vectors for assembly now fails on this branch.